core/src/test/java/com/linecorp/armeria/server/http/auth/AuthServiceTest.java

```diff
@@ -0,0 +1,129 @@
+package com.linecorp.armeria.server.http.auth;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertThat;
+
+import java.util.Base64;
+import java.util.Base64.Encoder;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Predicate;
+
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import com.google.common.collect.Maps;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponseWriter;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.http.AbstractHttpService;
+import com.linecorp.armeria.server.logging.LoggingService;
+
+public class AuthServiceTest {
+    private static final Encoder BASE_64_ENCODER = Base64.getEncoder();
+    private static final Server server;
+    private static final Predicate<Entry<String, String>> basicAuthPredicate =
+            new Predicate<Entry<String, String>>() {
+                private Map<String, Integer> counter = Maps.newHashMap();
+
+                @Override
+                public boolean test(Entry<String, String> entry) {
+                    String username = entry.getKey();
+                    String password = entry.getValue();
+
+                    if (!username.equals(password)) {
+                        return false;
+                    }
+
+                    if (counter.containsKey(username)) {
+                        int count = counter.get(username);
+
+                        if (count >= username.length()) {
+                            return false;
+                        } else {
+                            counter.put(username, count + 1);
```

Comment:
I think all this predicate logic is too complicated. Why don't you just use a `Map<String, String>` which contains username/password pairs?